task20/js/main.js

```diff
@@ -0,0 +1,93 @@
+import '../css/style.css'
+import { createElementWithClassName } from './utils/createElement'
+import { renderLoading } from './modules/loading'
+import { removeLoading } from './modules/loading'
+import { renderErrorMessage } from './modules/error'
+
+
+const parent = document.getElementById("js-parent");
+
+const getFetchData = async (endpoint) => {
+    const response = await fetch(endpoint);
+
+    if (!response.ok){
+        console.error(`$(response:status):$(response:statusText)`);
+        renderErrorMessage(parent, "問題が発生し表示することができません。");
+    }
+    return await response.json();
+};
+
+const fetchTableData = () => {
+    return new Promise((resolve) => {
+        setTimeout(() => {
+            resolve(getFetchData("https://mocki.io/v1/159d72ba-bd6b-429f-a28c-713eeaba9f72"));
+        }, 3000);
+        // setTimeout(() => {
+        //     resolve(getFetchData("https://mocki.io/v1/f12ae2d1-310d-4120-8749-47773d65e236"));
+        // }, 3000);//空配列
+    });
+};
+
+const initialize = async () => {
+    renderLoading(parent);
+    let userContentsData
+    try {
+        const json = await fetchTableData();
+        userContentsData = json.data;
+        if(userContentsData.length === 0){
+            renderErrorMessage(parent, "データがありません。");
+        }
+    } catch(error) {
+        renderErrorMessage(parent, "問題が発生し表示することができません。");
+    } finally {
+        removeLoading(parent);
+    }
+    renderTableContents(userTableColumn, userContentsData);
+};
+initialize();
```

Comment:
Q.

The following error appeared in the console when a 503 error or an empty array was passed.
![Screenshot at Nov 06 22-19-42](https://user-images.githubusercontent.com/67426438/200174221-bd1072c6-4d92-4e85-8174-ddeaab69c0f2.png)

I think this is because `parent.textContent` is done when displaying error messages,
so the child elements for loading are erased.

↓

After eliminating these errors, the renderTableContents() function would be executed even in the case of a 503 error or an empty array, which was not desirable.
What do you think?

---
I wrote the code to resolve.(Variable names are tentative.)
Please let me know if I got it wrong🙇‍♀️
```suggestion
const getUserData = async () => {
  renderLoading(parent);
  try {
    const json = await fetchTableData();
    const data = json.data;
    if (data.length === 0) {
      renderErrorMessage(hoge, "データがありません。");
      return;
    }
    return data;
  } catch (error) {
    renderErrorMessage(hoge, "問題が発生し表示することができません。");
  } finally {
    removeLoading(parent);
  }
};

const init = async () => {
  const data = await getUserData();
  data && renderTableContents(userTableColumn, data);
};
init();

```

Reply:
Thank you for all your thoughts.
[be5fbe7](https://github.com/seeeen044/JavaScript_handson/pull/21/commits/be5fbe7d3a3c2c5f21a4caa2f5ec642a700f65cb)

Reply:
修正ありがとうございます✨

日本語ですみません🙇‍♀️
ブラウザのコンソールでsyntaxエラーが出ていました。
![Screenshot at Nov 07 19-30-02](https://user-images.githubusercontent.com/67426438/200288457-050c29f0-573e-4796-95ed-fd302d9a453a.png)

調べたところ、fetchの503エラーのところが
```
setTimeout((resolve) => {
  getFetchData("https://httpstat.us/503")
  }, 3000);
```
となっていましたが、こうかと思いました。
```
setTimeout(() => {
      resolve(getFetchData("https://httpstat.us/503"));
    }, 3000);
```

そうすると、
503エラーの時、やはりremoveLoading()がうまく実行されていないエラーが出ていると思います。
対処すべきだと思ったのですがどうでしょうか..?
![Screenshot at Nov 07 19-22-06](https://user-images.githubusercontent.com/67426438/200287938-5308ad3f-bd11-4e08-a81d-4fb7fc307788.png)

Reply:
ご指摘ありがとうございました。ご教授いただいたコードを参考に修正いたしました。大変勉強になりました。エラーに関する認識が甘かったなと痛感しました⚡️
[f6233f1](https://github.com/seeeen044/JavaScript_handson/pull/21/commits/f6233f146d01156082d8d4f22da34ffeef477294)

Reply:
修正ありがとうございます！
503エラーの部分はOKだと思いました✨

1点、空配列を渡した時もrenderTableContents()が実行されてしまっているので、
表示が以下のようになっていました。

![Screenshot at Nov 10 20-52-53](https://user-images.githubusercontent.com/67426438/201084390-70fa51ca-846d-4416-916d-3a9b51888432.png)

if文にreturnが必要かと思いました。
```
  try {
    const json = await fetchTableData();
    const data = json.data;
    if (data.length === 0) {
      renderErrorMessage(parent, "データがありません。");
      return; // here!
    }
    return data;
```

Reply:
見落としておりました。ご指摘ありがとうございます！
何度も見て頂き本当にありがとうございます😭
修正しました。
[4d1ad5f](https://github.com/seeeen044/JavaScript_handson/pull/21/commits/4d1ad5fb6cbbc6fa4acce6153e28ac5cc92459f9)